### DISCUSSION TITLE
Should/could I be using NumPy?

### DISCUSSION THREAD

**Discussion by Bernmeister** [skyfielders/python-skyfield#800] [General]
I initially wrote my `Python` program calling `PyEphem` to compute the rise/set/azimuth/altitude for planets/minor planets/comets/stars/satellites.  The process is very simple, for example, for planets (in pseudocode):
```
for planet in planets
    compute rise/set/az/alt at a given date/time/lat/long/elev
```
Ditto for the other bodies.

Along comes `Skyfield` and I abstract so that I can switch to run either `PyEphem` or `Skyfield`.  At no time have I used `NumPy`; never needed it and was completely oblivious, particularly given `Skyfield` was rapidly becoming functionally closer `PyEphem`.  After reading this [comment](https://github.com/skyfielders/python-skyfield/pull/526#issuecomment-760389968) I wondered if there's a better way to write my code when calling `Skyfield`, particularly if I'm looping over bodies at the same date/time (and lat/long/elev).

Rather than pass each body (of the same type) one by one, instead, somehow, pass in a list of bodies (with or without the help of `NumPy`).  I have seen [mention](https://rhodesmill.org/skyfield/api-time.html) of using multiple date/times within a single `Time` object, but this is the opposite to what I want: multiple objects for a given date/time.

**Comment by brandon-rhodes** [comment]
I am not sure whether it would be easy to convince NumPy to run efficient operations over different bodies.

Skyfield can be efficient with an array of times because each planet has a big array of positions, and Skyfield can have NumPy produce 100 positions with no more Python code than producing 1 position—so the only additional expense of the 99 more positions is the floating point math that's kicked off, which is a fast set of operations when compared to running further Python code.

But two different planets have two different time-indexed arrays that aren't lined up next to each other in memory in a way that would make them easy to traverse—in fact they probably have different step sizes, since fast planets with tight orbits store more data points than slow planets with big lazy orbits.

But maybe there's some way I'm not thinking of that would let _n_ planets be pre-broadcast into some kind of common array that would then only need to be operated on once? I'm open to ideas, and I know that there are folks with much more NumPy experience!
